File: samples/AdventureWorksLT/ProductCategory.cs
﻿using DevZest.Data;
using DevZest.Data.SqlServer;
using DevZest.Data.Annotations;
using System.Threading;

namespace DevZest.Samples.AdventureWorksLT
{
    public class ProductCategory : BaseModel<ProductCategory.PK>
    {
        [DbPrimaryKey("PK_ProductCategory_ProductCategoryID", Description = "Primary key (clustered) constraint")]
        public sealed class PK : PrimaryKey
        {
            public PK(_Int32 productCategoryID)
                : base(productCategoryID)
            {
            }
        }

        public class Key : Key<PK>
        {
            static Key()
            {
                Register((Key _) => _.ProductCategoryID, _ProductCategoryID);
            }

            protected override PK CreatePrimaryKey()
            {
                return new PK(ProductCategoryID);
            }

            public _Int32 ProductCategoryID { get; private set; }
        }

        public class Ref : Ref<PK>
        {
            static Ref()
            {
                Register((Ref _) => _.ProductCategoryID, _ProductCategoryID);
            }

            public _Int32 ProductCategoryID { get; private set; }

            protected override PK CreateForeignKey()
            {
                return new PK(ProductCategoryID);
            }
        }

        public class Lookup : Projection
        {
            static Lookup()
            {
                Register((Lookup _) => _.Name, _Name);
            }

            public _String Name { get; private set; }
        }

        public static readonly Mounter<_Int32> _ProductCategoryID = RegisterColumn((ProductCategory _) => _.ProductCategoryID);
        public static readonly Mounter<_Int32> _ParentProductCategoryID = RegisterColumn((ProductCategory _) => _.ParentProductCategoryID);
        public static readonly Mounter<_String> _Name = RegisterColumn((ProductCategory _) => _.Name);

        static ProductCategory()
        {
            RegisterChildModel((ProductCategory x) => x.SubCategories, (ProductCategory x) => x.FK_ParentProductCategory);
        }

        public ProductCategory SubCategories { get; private set; }

        protected sealed override PK CreatePrimaryKey()
        {
            return new PK(ProductCategoryID);
        }

        [Identity(1, 1)]
        [DbColumn(Description = "Primary key for ProductCategory records.")]
        public _Int32 ProductCategoryID { get; private set; }

        [DbColumn(Description = "Product category identification number of immediate ancestor category. Foreign key to ProductCategory.ProductCategoryID.")]
        public _Int32 ParentProductCategoryID { get; private set; }

        private PK _fk_parentProductCategory;
        public PK FK_ParentProductCategory
        {
            get { return _fk_parentProductCategory ?? (_fk_parentProductCategory = new PK(ParentProductCategoryID)); }
        }

        [UdtName]
        [DbColumn(Description = "Category description.")]
        [Required]
        [AsNVarChar(50)]
        [Unique(Name = "AK_ProductCategory_Name", Description = "Unique nonclustered constraint.")]
        public _String Name { get; private set; }
    }
}

File: samples/AdventureWorksLT.VB/ProductCategory.vb
Imports DevZest.Data
Imports DevZest.Data.SqlServer
Imports DevZest.Data.Annotations

Namespace DevZest.Samples.AdventureWorksLT
    Public Class ProductCategory
        Inherits BaseModel(Of ProductCategory.PK)

        <DbPrimaryKey("PK_ProductCategory_ProductCategoryID", Description:="Primary key (clustered) constraint")>
        Public NotInheritable Class PK
            Inherits PrimaryKey

            Public Shared Function ValueOf(productCategoryID As Integer) As IDataValues
                Return DataValues.Create(_Int32.Const(productCategoryID))
            End Function

            Public Sub New(ByVal productCategoryID As _Int32)
                MyBase.New(productCategoryID)
            End Sub

            Public ReadOnly Property ProductCategoryID As _Int32
                Get
                    Return GetColumn(Of _Int32)(0)
                End Get
            End Property
        End Class

        Public Class Key
            Inherits Key(Of PK)

            Shared Sub New()
                Register(Function(x As Key) x.ProductCategoryID, _ProductCategoryID)
            End Sub

            Protected Overrides Function GetPrimaryKey() As PK
                Return New PK(ProductCategoryID)
            End Function

            Private m_ProductCategoryID As _Int32
            Public Property ProductCategoryID As _Int32
                Get
                    Return m_ProductCategoryID
                End Get
                Private Set
                    m_ProductCategoryID = Value
                End Set
            End Property
        End Class

        Public Class Ref
            Inherits Ref(Of PK)

            Shared Sub New()
                Register(Function(x As Ref) x.ProductCategoryID, _ProductCategoryID)
            End Sub

            Private m_ProductCategoryID As _Int32
            Public Property ProductCategoryID As _Int32
                Get
                    Return m_ProductCategoryID
                End Get
                Private Set
                    m_ProductCategoryID = Value
                End Set
            End Property

            Protected Overrides Function GetForeignKey() As PK
                Return New PK(ProductCategoryID)
            End Function
        End Class

        Public Class Lookup
            Inherits Projection

            Shared Sub New()
                Register(Function(x As Lookup) x.Name, _Name)
            End Sub

            Private m_Name As _String
            Public Property Name As _String
                Get
                    Return m_Name
                End Get
                Private Set
                    m_Name = Value
                End Set
            End Property
        End Class

        Public Shared ReadOnly _ProductCategoryID As Mounter(Of _Int32) = RegisterColumn(Function(x As ProductCategory) x.ProductCategoryID)
        Public Shared ReadOnly _ParentProductCategoryID As Mounter(Of _Int32) = RegisterColumn(Function(x As ProductCategory) x.ParentProductCategoryID)
        Public Shared ReadOnly _Name As Mounter(Of _String) = RegisterColumn(Function(x As ProductCategory) x.Name)

        Shared Sub New()
            RegisterChildModel(Function(x As ProductCategory) x.SubCategories, Function(x As ProductCategory) x.FK_ParentProductCategory)
        End Sub

        Private m_SubCategories As ProductCategory
        Public Property SubCategories As ProductCategory
            Get
                Return m_SubCategories
            End Get
            Private Set
                m_SubCategories = Value
            End Set
        End Property

        Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
            Return New PK(ProductCategoryID)
        End Function

        Private m_ProductCategoryID As _Int32
        <Identity(1, 1)>
        <DbColumn(Description:="Primary key for ProductCategory records.")>
        Public Property ProductCategoryID As _Int32
            Get
                Return m_ProductCategoryID
            End Get
            Private Set
                m_ProductCategoryID = Value
            End Set
        End Property

        Private m_ParentProductCategoryID As _Int32
        <DbColumn(Description:="Product category identification number of immediate ancestor category. Foreign key to ProductCategory.ProductCategoryID.")>
        Public Property ParentProductCategoryID As _Int32
            Get
                Return m_ParentProductCategoryID
            End Get
            Private Set
                m_ParentProductCategoryID = Value
            End Set
        End Property

        Private m_FK_productCategory As PK
        <DbColumn(Description:="Category description.")>
        Public ReadOnly Property FK_ParentProductCategory As PK
            Get
                If m_FK_productCategory Is Nothing Then m_FK_productCategory = New PK(ParentProductCategoryID)
                Return m_FK_productCategory
            End Get
        End Property

        Private m_Name As _String
        <UdtName>
        <Required>
        <AsNVarChar(50)>
        <Unique(Name:="AK_ProductCategory_Name", Description:="Unique nonclustered constraint.")>
        Public Property Name As _String
            Get
                Return m_Name
            End Get
            Private Set
                m_Name = Value
            End Set
        End Property
    End Class
End Namespace
